### PR TITLE
nxp: k6x: move clk divider options to device tree

### PR DESCRIPTION
**Pull request by aunsbjerg**
With inspiration from #26610, I've started moving the clock divider settings for the SIM module from the SoC's kconfig to device tree. The defaults are now set in `nxp_k6x.dtsi` and can be overwritten by boards as needed.

### CODE REVIEW

**Comment by henrikbrixandersen** [comment]
@aunsbjerg You need to fix the compliance issues with the missing bodies in the Git commit messages.

**Review by galak — CHANGES_REQUESTED** [review]
So I think this should be more inline with clock devicetree bindings:

https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/clock/fixed-factor-clock.yaml

So something like:

```
core_clk: core_clk {
        #clock-cells = <0>;
        compatible = "fixed-factor-clock";
        clocks = <&mcgoutclk>;
        clock-div = <1>;
};
```

File: soc/arm/nxp_kinetis/k6x/soc.h [hunk 1]
```diff
@@ -24,8 +24,12 @@
 #elif defined(CONFIG_SOC_MK66F18)
 #define SYSCLK_DEFAULT_IOSC_HZ MHZ(180)
 #endif
+
+#define CLOCK_DIVIDER(prop) \
+	DT_PROP_OR(DT_INST(0, nxp_kinetis_sim), prop, 1) - 1
+
 #define BUSCLK_DEFAULT_IOSC_HZ (SYSCLK_DEFAULT_IOSC_HZ / \
-				CONFIG_K6X_BUS_CLOCK_DIVIDER)
+				CLOCK_DIVIDER(bus_clk_divider))
```
Comment (Kludentwo) [inline soc/arm/nxp_kinetis/k6x/soc.h:32]: CONFIG_K6X_BUS_CLOCK_DIVIDER was 2 but using the macro makes it 1?

**Comment by aunsbjerg** [comment]
> So I think this should be more inline with clock devicetree bindings:
> 
> https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/clock/fixed-factor-clock.yaml
> 
> So something like:
> 
> ```
> core_clk: core_clk {
>         #clock-cells = <0>;
>         compatible = "fixed-factor-clock";
>         clocks = <&mcgoutclk>;
>         clock-div = <1>;
> };
> ```

I added a `fixed-factor-clock` binding as suggested and end up with something like this:

```
// nxp_k6x.dtsi
sim: sim@40047000 {
	compatible = "nxp,kinetis-sim";
	reg = <0x40047000 0x1060>;
	label = "SIM";
	#clock-cells = <3>;

	core_clk: core_clk {
		compatible = "fixed-factor-clock";
		clocks = <&mcg KINETIS_MCG_FIXED_FREQ_CLK>; // need to add proper clock name
		clock-div = <1>;
		#clock-cells = <0>;
	}; 
};
```
And:
```
// k6x soc.c
#define CLOCK_DIVIDER(clk) \
	DT_PROP_OR(DT_NODELABEL(clk), clock_div, 1) - 1

static const sim_clock_config_t simConfig = {
	.pllFllSel = DT_PROP(DT_INST(0, nxp_kinetis_sim), pllfll_select),
	.er32kSrc = DT_PROP(DT_INST(0, nxp_kinetis_sim), er32k_select),
	.clkdiv1 = SIM_CLKDIV1_OUTDIV1(CLOCK_DIVIDER(core_clk)) |
		   SIM_CLKDIV1_OUTDIV2(CLOCK_DIVIDER(bus_clk)) |
		   SIM_CLKDIV1_OUTDIV3(CLOCK_DIVIDER(flexbus_clk)) |
		   SIM_CLKDIV1_OUTDIV4(CLOCK_DIVIDER(flash_clk)),
};
```

However, I'm not sure if this is what you are suggesting or if this is the way to go. This differs quite a bit from the `nxp,kinetis-scg.yaml` binding + driver for the `ke1xf` SoC that has most (if not all) clock settings available in device tree. Thoughts?

**Comment by galak** [comment]
> However, I'm not sure if this is what you are suggesting or if this is the way to go. This differs quite a bit from the `nxp,kinetis-scg.yaml` binding + driver for the `ke1xf` SoC that has most (if not all) clock settings available in device tree. Thoughts?

What you've suggested is exactly inline with what I'm thinking and asking for.  The support for the `ke1xf` should eventually be changed over to look like what you've done.  Its a historic example that we should try and avoid propagating and fix as we can.

The reason is over time we'll want to have proper clock tree's in devicetree to handle power management correctly.  At the level of support we've had historically this wasn't as large of a concern.  However as we keep improving things this is the direction we should head.

**Comment by aunsbjerg** [comment]
> > However, I'm not sure if this is what you are suggesting or if this is the way to go. This differs quite a bit from the `nxp,kinetis-scg.yaml` binding + driver for the `ke1xf` SoC that has most (if not all) clock settings available in device tree. Thoughts?
> 
> What you've suggested is exactly inline with what I'm thinking and asking for. The support for the `ke1xf` should eventually be changed over to look like what you've done. Its a historic example that we should try and avoid propagating and fix as we can.
> 
> The reason is over time we'll want to have proper clock tree's in devicetree to handle power management correctly. At the level of support we've had historically this wasn't as large of a concern. However as we keep improving things this is the direction we should head.

Super, that makes sense. The more I looked at the existing code, the more confused I got.
I've pushed some changes that I believe implements what you are suggesting.

File: dts/bindings/clock/fixed-rate-clock.yaml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+description: Generic fixed factor clock provider
+
+compatible: "fixed-factor-clock"
+
+include: clock-controller.yaml
+
+properties:
+    label:
+      type: string
```
Comment (galak) [inline dts/bindings/clock/fixed-rate-clock.yaml:12]: Remove `label` its not needed here.

**Review by galak — COMMENTED** [review]
LGTM.  One change on dropping `label` from the binding.

Can you look at making this change to the other relevant Kinetis SoCs:

```
soc/arm/nxp_kinetis/k2x/Kconfig.soc:config K22_FLASH_CLOCK_DIVIDER
soc/arm/nxp_kinetis/k2x/soc.c:             SIM_CLKDIV1_OUTDIV4(CONFIG_K22_FLASH_CLOCK_DIVIDER - 1),
soc/arm/nxp_kinetis/k6x/Kconfig.soc:config K6X_FLASH_CLOCK_DIVIDER
soc/arm/nxp_kinetis/k6x/soc.c:             SIM_CLKDIV1_OUTDIV4(CONFIG_K6X_FLASH_CLOCK_DIVIDER - 1),
soc/arm/nxp_kinetis/k8x/Kconfig.soc:config K8X_FLASH_CLOCK_DIVIDER
soc/arm/nxp_kinetis/k8x/soc.c:             SIM_CLKDIV1_OUTDIV4(CONFIG_K8X_FLASH_CLOCK_DIVIDER - 1),
soc/arm/nxp_kinetis/kv5x/Kconfig.soc:config KV5X_FLASH_CLOCK_DIVIDER
soc/arm/nxp_kinetis/kv5x/soc.c:            SIM_CLKDIV1_OUTDIV4(CONFIG_KV5X_FLASH_CLOCK_DIVIDER - 1),
soc/arm/nxp_kinetis/kwx/Kconfig.soc:config KW2XD_FLASH_CLOCK_DIVIDER
soc/arm/nxp_kinetis/kwx/soc_kw2xd.c:               SIM_CLKDIV1_OUTDIV4(CONFIG_KW2XD_FLASH_CLOCK_DIVIDER - 1),
```

**Comment by aunsbjerg** [comment]
> LGTM. One change on dropping `label` from the binding.
> 
> Can you look at making this change to the other relevant Kinetis SoCs:
> 
> ```
> soc/arm/nxp_kinetis/k2x/Kconfig.soc:config K22_FLASH_CLOCK_DIVIDER
> soc/arm/nxp_kinetis/k2x/soc.c:             SIM_CLKDIV1_OUTDIV4(CONFIG_K22_FLASH_CLOCK_DIVIDER - 1),
> soc/arm/nxp_kinetis/k6x/Kconfig.soc:config K6X_FLASH_CLOCK_DIVIDER
> soc/arm/nxp_kinetis/k6x/soc.c:             SIM_CLKDIV1_OUTDIV4(CONFIG_K6X_FLASH_CLOCK_DIVIDER - 1),
> soc/arm/nxp_kinetis/k8x/Kconfig.soc:config K8X_FLASH_CLOCK_DIVIDER
> soc/arm/nxp_kinetis/k8x/soc.c:             SIM_CLKDIV1_OUTDIV4(CONFIG_K8X_FLASH_CLOCK_DIVIDER - 1),
> soc/arm/nxp_kinetis/kv5x/Kconfig.soc:config KV5X_FLASH_CLOCK_DIVIDER
> soc/arm/nxp_kinetis/kv5x/soc.c:            SIM_CLKDIV1_OUTDIV4(CONFIG_KV5X_FLASH_CLOCK_DIVIDER - 1),
> soc/arm/nxp_kinetis/kwx/Kconfig.soc:config KW2XD_FLASH_CLOCK_DIVIDER
> soc/arm/nxp_kinetis/kwx/soc_kw2xd.c:               SIM_CLKDIV1_OUTDIV4(CONFIG_KW2XD_FLASH_CLOCK_DIVIDER - 1),
> ```

Done. Let me know if you prefer to have some of the commits squashed.
Also, b82b656 removed some unused defines, which is not strictly part of this PR. 
The blinky sample still works on frdm_k64f and frdm_k22f boards - don't have boards to test the other SoC changes.

**Comment by aunsbjerg** [comment]
Buildkite errors should be fixed now.

The issues were caused by a missing mcg node in the kl25z board and the kw40z, kw41z and kv5x device tree using undefined bindings for the mcg instead of using "nxp,kinetis-mcg".

**Review by MaureenHelm — APPROVED** [review]
This looks great, thank you @aunsbjerg ! We've needed to start moving clocks to dts for a while now.

**Comment by galak** [comment]
@aunsbjerg any reason for this to still be labeled DNM?

**Review by galak — APPROVED** [review]

**Comment by aunsbjerg** [comment]
> @aunsbjerg any reason for this to still be labeled DNM?

No, just me not paying attention. Fixed.

**Review by henrikbrixandersen — APPROVED** [review]
Looks good, thank you.

**Pull request merged**